### PR TITLE
Feature/shahafandlika/co classifier

### PR DESCRIPTION
**Pull request by LikaLiel**
Updated WeightedLogisticRegression class

### CODE REVIEW

File: assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.pipeline import make_pipeline
+# from pagayapro.paths.data_paths import ASSIGNMENT4_DATA
+# import os
+import numpy_financial as npf
+from sklearn import preprocessing
+
+class WeightedLogisticregression():
+    THRESHOLD = 0.45
```
Comment (shai-shechter) [inline assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:12]: please move the constants to be parameters of the class constructor, with default parameters 0.45, {1:5, 0:1} and 0.4

File: assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        :return: a fitted model
+        """
+        preprocessed_X = self.__apply_transformation_to_data(X, y)
+        self.weighted_pipe = make_pipeline(StandardScaler(), LogisticRegression(class_weight=WeightedLogisticregression.WEIGHTS))
+        self.weighted_pipe.fit(preprocessed_X, self.y)
+        return self
+
+    def predict(self, X):
+        """
+        Must be called after fit. predicts the
```
Comment (shai-shechter) [inline assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:41]: finish this partial doc please

File: assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        :param X:
+        :return: the loans the model thinks are good. 1 is will Co, 0 is wont Co
+        """
+        preprocessed_X = self.__apply_transformation_to_data(X, None)
+        pipe_proba = self.weighted_pipe.predict_proba(preprocessed_X)[:1]
+        return (pipe_proba <= WeightedLogisticregression.THRESHOLD).astype('int')
```
Comment (shai-shechter) [inline assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:47]: Are you sure about this? This will return 1 if the loan has a probability <= threshold to CO, is this what you want?

Comment (LikaLiel) [reply assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:47]: it is a bit confusing but self.weighted_pipe.predict_proba(preprocessed_X)[:,1] is the probability to belong to label 0 (ie not-CO) so this is why it works
if we did self.weighted_pipe.predict_proba(preprocessed_X)[:,0] we would have >= threshold

File: assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            if y.isna().sum() == 0:
+                self.y = y
+            else:
+                self.y = (~y.isna()).astype('int')  # 1 is CO, 0 no CO
+            self.y.columns = ['co_mob']
+            # 1) I need to add a סיומת to the y column
+            # so that it will certainly be unique, but this implementation should be good enough for this demo
+            # 2) if i was making this app generic i wouldn't demand "occupation"
+            # i would simply convert all catergorical data this way
+            if 'occupation' in X.columns:
```
[the remark this answers is not in the document]
Comment (LikaLiel) [reply assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:75]: i dont know how generic to make this class. if i make it fully generic i never need this, but when i want to make it do what we did in the notebook i need to convert the occupation column to a numeric value rather than simply drop it..

File: assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                self.occupation_map = pd.concat([X, self.y], axis=1).groupby('occupation')['co_mob'].mean()
+
+        numerical_X = self.__make_numerical(X)
+        if self.good_columns is None:
```
[the remark this answers is not in the document]
Comment (LikaLiel) [reply assignments/assignment4/models/shahafandlika/WeightedLogisticRegression.py:83]: This gets rid of columns that are not numeric if there were any
